### PR TITLE
Feat/monitoring and alerting

### PR DESCRIPTION
**Pull request by jusdino**
### Description List
- Add alert SNS topic for monitoring/alerting in each environment
- Instrument Lambda, SQS, API Gateway, CodePipeline alerts
- Instrument email, Slack notifications

### Testing List
- Code review

Closes #
#94

### CODE REVIEW

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: backend/compact-connect/pipeline/backend_pipeline.py [hunk 1]
```diff
@@ -98,3 +105,22 @@ def build_pipeline(self) -> None:
...
+
+    def _add_alarms(self):
+        NotificationRule(
+            self, 'NotificationRule',
+            source=self.pipeline,
+            events=[
+                'codepipeline-pipeline-pipeline-execution-started',
+                'codepipeline-pipeline-pipeline-execution-failed',
+                'codepipeline-pipeline-pipeline-execution-succeeded',
+                'codepipeline-pipeline-manual-approval-needed'
```
Comment (jlkravitz) [inline backend/compact-connect/pipeline/backend_pipeline.py:119]: What's the naming scheme here? (Trying to parse why `pipeline` is written out three times :D

File: backend/compact-connect/stacks/persistent_stack/staff_users.py [hunk 2]
```diff
@@ -33,7 +34,7 @@ def __init__(
             removal_policy=removal_policy,
             **kwargs
         )
-        stack: AppStack = AppStack.of(self)
+        stack: ps.PersistentStack = ps.PersistentStack.of(self)
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/persistent_stack/staff_users.py:37]: What changed here? Why was `AppStack` previously correct?

File: backend/compact-connect/pipeline/__init__.py [hunk 3]
```diff
@@ -51,19 +52,30 @@ def __init__(
...
+        )
+
         access_logs_bucket = AccessLogsBucket(
             self, 'AccessLogsBucket',
             removal_policy=removal_policy,
             auto_delete_objects=removal_policy == RemovalPolicy.DESTROY
         )
 
+        # Allows us to override the default branching scheme for the test environment, via context variable
+        pre_prod_trigger_branch = pipeline_environment_context.get('pre_prod_trigger_branch', 'development')
```
Comment (jlkravitz) [inline backend/compact-connect/pipeline/__init__.py:70]: thanks for doing this!

File: backend/compact-connect/common_constructs/slack_channel_configuration.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            suppressions=[
+                {
+                    'id': 'AwsSolutions-IAM4',
+                    'appliesTo': ['Policy::arn:<AWS::Partition>:iam::aws:policy/job-function/ViewOnlyAccess'],
+                    'reason': 'This role is general-purpose for operations integration and the AWS-managed '
+                              'ViewOnlyAccess policy is suitable'
+                },
+                {
+                    'id': 'AwsSolutions-IAM5',
+                    'reason': 'This role is intended to be able to query logs across teh account to facilitate '
```
Comment (jlkravitz) [inline backend/compact-connect/common_constructs/slack_channel_configuration.py:45]:
```suggestion
                    'reason': 'This role is intended to be able to query logs across the account to facilitate '
```

**Review by jlkravitz — COMMENTED** [review]
Looks great! A couple questions but otherwise looking good.

File: backend/compact-connect/README.md [hunk 5]
```diff
@@ -134,6 +134,14 @@ that is done, perform the following steps to deploy the CI/CD pipeline into the
 - Copy the `cdk.context.production-example.json` file to `cdk.context.json` and update accounts and other identifiers,
   including the Code Star connection you just had created to match the identifiers for your actual accounts and
   resources.
+- Optional: If a Slack integration is desired for operational support, have someone with permission to install Slack
+  apps in your workspace and Admin access to each of the Test, Prod, and Deploy accounts log into the Test AWS account
+  and go to the Chatbot service. Select 'Slack' under the **Configure a chat client** box and click **Configure
+  client**, then follow the Slack authorization prompts. This will authorize AWS to integrate with the channels you
+  identify in your `cdk.context.json` file. Repeat this process for the Prod and Deploy accounts as well. For each
+  Slack channel you want to integrate, be sure to invite your new AWS app to those channels. Update the
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/README.md:142]: You just `@aws`, just like anybody else 😄

File: backend/compact-connect/pipeline/backend_pipeline.py (hunk 1 above)
Comment (jusdino) [reply backend/compact-connect/pipeline/backend_pipeline.py:119]: These are just the [names that CodePIpeline defines](https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#events-ref-pipeline)

File: backend/compact-connect/stacks/persistent_stack/staff_users.py (hunk 2 above)
Comment (jusdino) [reply backend/compact-connect/stacks/persistent_stack/staff_users.py:37]: `AppStack` is the base class of `PersistentStack`. We're using a more specific attribute (`PersistentStack.alarm_topic`), so I'm including a more specific type hint.

File: backend/compact-connect/common_constructs/alarm_topic.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        )
+
+    def _configure_s3_principal(self, master_key: IKey):
+        stack = Stack.of(self)
+        s3_principal = ServicePrincipal('s3.amazonaws.com')
+
+        master_key.grant_encrypt_decrypt(s3_principal)
+
+        self.add_to_resource_policy(
+            # Allow S3 to publish events to this topic, but only from this account
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/common_constructs/alarm_topic.py:83]: It's just a matter of how AWS works differently between S3 events and CloudWatch events and what parameters are available to control requests between the two. "this" account is the AWS account this service is deployed to.

[the remark this answers is not in the document]
Comment (jlkravitz) [reply backend/compact-connect/common_constructs/alarm_topic.py:83]: Got it thanks!

File: backend/compact-connect/README.md
Comment (jlkravitz) [inline backend/compact-connect/README.md:1]: Can any of this be tested? Seems like it'd be difficult (and probably not worth it) to do so, but I gotta ask :)

Comment (jusdino) [reply backend/compact-connect/README.md:1]: Yeah, it's pretty tough to automatically test these. I can demo some of the alerts by manually forcing an 'error' condition in the app.

Comment (jlkravitz) [reply backend/compact-connect/README.md:1]: makes sense!

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen This is good to go.

**Pull request merged**